dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/event/WorkflowStartEventHandler.java

```diff
@@ -66,11 +68,35 @@ public void handleWorkflowEvent(final WorkflowEvent workflowEvent) throws Workfl
                         if (processInstance.getTimeout() > 0) {
                             stateWheelExecuteThread.addProcess4TimeoutCheck(processInstance);
                         }
+                    } else if (WorkflowSubmitStatue.ERROR == workflowSubmitStatue) {
+                        log.error(
+                                "Failed to submit the workflow instance, will resend the workflow start event: {}, and reduce submit times that can be used",
+                                workflowEvent);
+                        workflowEvent.setMaxSubmitTimes(workflowEvent.getMaxSubmitTimes() - 1);
+                        if (workflowEvent.getMaxSubmitTimes() >= 0) {
+                            workflowEventQueue.addEvent(workflowEvent);
+                        } else {
+                            WorkflowStateEvent stateEvent = WorkflowStateEvent.builder()
+                                    .processInstanceId(processInstance.getId())
+                                    .type(StateEventType.PROCESS_SUBMIT_FAILED)
+                                    .status(WorkflowExecutionStatus.FAILURE)
+                                    .build();
+                            workflowExecuteRunnable.addStateEvent(stateEvent);
+                        }
                     } else {
                         // submit failed will resend the event to workflow event queue
                         log.error("Failed to submit the workflow instance, will resend the workflow start event: {}",
                                 workflowEvent);
-                        workflowEventQueue.addEvent(workflowEvent);
```

Comment:
I think it's better to set the workflow instance status to failed if the submit unsuccessful, without retrying. It is up to the user to decide whether to manually retry. WDYT?

Reply:
If the database is unstable and the connection times out, or other similar problems, add the number of retries to avoid user actions